### PR TITLE
Restrict certain permissions by subject type

### PR DESCRIPTION
**Pull request by FrederikBolding**
## Description

Adds a `subjectTypes` field to permission specifications, allowing certain permissions to be limited to access from certain subject types. 

If a request for permissions is made from a non-matching subject type we throw a method or endowment not found error.

## Changes

- **Added**: Added `subjectTypes` field to permission specifications
   - This field can be used to limit which subject types can request a given permission

## Checklist

- [x] I've updated the test suite for new or updated code as appropriate
- [x] I've updated documentation for new or updated code as appropriate (note: this will usually be JSDoc)
- [x] I've highlighted breaking changes using the "BREAKING" category above as appropriate

### CODE REVIEW

File: packages/permission-controller/src/PermissionController.ts [hunk 1]
```diff
@@ -853,6 +855,25 @@ export class PermissionController<
       throw failureError;
     }
 
+    if (specification.subjectTypes) {
```
Comment (Mrtenz) [inline packages/permission-controller/src/PermissionController.ts:858]:
```suggestion
    if (specification.subjectTypes?.length > 0) {
```

**Review by Mrtenz — CHANGES_REQUESTED** [review]

File: packages/permission-controller/src/PermissionController.ts (hunk 1 above)
Comment (FrederikBolding) [reply packages/permission-controller/src/PermissionController.ts:858]: `'specification.subjectTypes.length' is possibly 'undefined'`

Comment (FrederikBolding) [reply packages/permission-controller/src/PermissionController.ts]: Should be addressed now

**Review by Mrtenz — APPROVED** [review]

File: packages/permission-controller/src/PermissionController.ts [hunk 2]
```diff
@@ -1717,6 +1719,27 @@ export class PermissionController<
     },
   ): void {
     const { allowedCaveats, validator } = specification;
+
+    if (
+      specification.subjectTypes?.length &&
+      specification.subjectTypes.length > 0
```
Comment (GuillaumeRx) [inline packages/permission-controller/src/PermissionController.ts:1725]: Can't we just do `specification.subjectType?.length > 0` ?

**Review by GuillaumeRx — DISMISSED** [review]

File: packages/permission-controller/src/PermissionController.ts [hunk 3]
```diff
@@ -1717,6 +1719,27 @@ export class PermissionController<
     },
   ): void {
     const { allowedCaveats, validator } = specification;
+
+    if (
```
Comment (Gudahtt) [inline packages/permission-controller/src/PermissionController.ts:1723]: Nit: In the JSDoc comment for this method, there is a list of things that it's responsible for validating. We should add an entry for the subject type

**Review by Gudahtt — APPROVED** [review]

**Pull request merged**